docs/IWYUStyle.md

```diff
@@ -0,0 +1,242 @@
+* Pointer asterisk and reference ampersand binds to the type name, not the
+  declared variable (i.e. declarations look like `Foo* f;` and `Bar& b;`)
+* When in doubt, mimic the code around you.
+
+IWYU has a `.clang-format` file, so everything should play out nicely if you use
+clang-format on your edited lines (but PLEASE do not reformat more than lines
+you touch, especially not whole files).
+
+We observe a few [LLVM-inspired rules][1] for IWYU:
+
+* **Use 'auto' sparingly** -- we don't subscribe to the "Almost always auto"
+guideline in IWYU. In line with LLVM tradition we use `auto` when it improves
+readability, and take that to mean when the type is already clearly visible on
+the line (e.g. `const auto& d = cast<Derived>();`) or when iterator declarations
+would grow too unwieldy (e.g. `const auto i = some_map.begin();`).
+
+* **Prefer 'static' to anonymous namespaces** -- while `static` is decidedly a
+C-ism, we find it's a readability win compared to anonymous namespaces to mark
+functions for internal linkage. The LLVM coding standards makes the argument
+well, so no use to repeat it here.
+
+Other than that, we mostly follow mainstream C++ style, if there is such a
+thing.
+
+
+## Python ##
+
+Older IWYU Python code follows the [Google convention for Python code][3], but
+newer scripts typically just use [PEP8][4].
+
+Again, mimic the code around you.
+
+
+## Regression tests ##
+
+IWYU uses a regression testing framework purpose-built to verify IWYU
+diagnostics.
+
+There are a few conventions for IWYU testcases we use to keep the test suite
+maintainable and effective. Note that we have many tests breaking these
+guidelines, but that's considered a bug that we will work on over time.
+
+### Testcase per construct ###
+
+IWYU started out with one single testcase -- `badinc.cc` -- to cover all aspects
+of IWYU. This has proved hard to maintain, as many assertions are subtly
+interconnected and it can be difficult to map a test failure back to the feature
+that triggered it.
+
+Now that IWYU handles most of the basics, we try to add new testcases organized
+by the language construct or IWYU feature they cover. This tends to produce a
+large number of testcases, and we'll probably have to go back and organize them
+further at some point.
+
+Some things to keep in mind:
+
+* Keep testcase filenames short -- this is a case of "as short as possible, but
+  no shorter". Every file has its name in the license header, the stem is
+  repeated for supporting headers, and testcase names occasionally need to be
+  typed without completion, so it's nice if they're short and sweet. But they
+  also need to be meaningful. Find the balance.
+* If a change pertains to a feature covered by an existing testcase, consider
+  extending it instead of creating a new one.
+* Don't use another testcase's supporting headers; that creates unwanted
+  coupling between tests. But see below for a set of headers we do encourage
+  sharing.
```

Comment:
Something I always try to do is verify if the test case that I created fails before applying the changes. Otherwise, chances are that I am not testing what I expected :)